### PR TITLE
Code editor

### PR DESCRIPTION
**Pull request by cchan987**
Adding the ace library and setting it up on server and client size
-Toggle button to open/close editor
-Editor dynamically updates the netgraph
-Editor gets the source code for the model that you are running and displays it to the user
-Editor polls for changes every 100ms

### CODE REVIEW

**Comment by jgosmann** [comment]
btw: [You should setup your git author email correctly.](https://help.github.com/articles/setting-your-email-in-git/)

**Comment by jgosmann** [comment]
PR addresses #273.

**Comment by jgosmann** [comment]
Is this ready for review or still WIP?

**Comment by tcstewar** [comment]
still WIP  :)

**Comment by tcstewar** [comment]
There's now decent handling of errors (showing the exception in the code editor) in https://github.com/nengo/nengo_viz/tree/code_editor_errors

**Comment by tcstewar** [comment]
Also, lots of extra files for ace seem to be here.  The only ones I think are necessary are ace.js and mode-python.js

**Comment by cchan987** [comment]
This is ready to be looked at for merge

**Comment by celiasmith** [comment]
Nice!  I like the white background more than i thought :smile: .  I'm assuming you want interaction comments as well as look & feel comments.
Interaction (starting with 2d.py):
1. When I typed an ensemble, all the connections (unrelated to what I was doing) disappeared.
[NameError: name 'e' is not defined caused it... I was typing "ens2"]
2. When I copy pasted an ensemble and renamed it, it didn't get created
3. There were lots of errors thrown, so I suspected it stopped working while doing that.

Feel:
1. There's no cursor indication I can drag to change the width... often at least.
2. When an ensemble was created, it was tiny.... probably because of my zoom at the time or something?

**Comment by cchan987** [comment]
> 1. When I typed an ensemble, all the connections (unrelated to what I was doing) disappeared.
> [NameError: name 'e' is not defined caused it... I was typing "ens2"]
> 2. When I copy pasted an ensemble and renamed it, it didn't get created
> 3. There were lots of errors thrown, so I suspected it stopped working while doing that.

I know Terry worked on some error handling for the editor in another branch, this will hopefully mitigate the errors you are having once they are merged

**Comment by jgosmann** [comment]
I get the `default` cursor in Firefox over the text instead of this selection bar. Maybe that's fixed with PR #284?

**Comment by jgosmann** [comment]
Will do further testing later.

**Comment by jgosmann** [comment]
Can we make the row highlight the same color as in the highlight in the time bar? (Or at least not the same gray as everywhere else. Looks strange if the cursor is in the first row.)

The editor covers the shadow of the top toolbar (but not the bottom toolbar). I think, neither shadow should be covered.

The gray behind the line (#efefef) numbers slightly differs from the toolbar gray (#eeeeee). And yes, that's just barely enough for me to see it where the grays touch. :eyeglasses: Though, I verified it with a digital pipette.

It would be cool to be able to create a new file from within the GUI.

How do I save anything?

(I want a different color scheme for syntax highlighting. But that's of minor importance right now.)

**Comment by jgosmann** [comment]
If I type to add ensemble, stuff I'm not editing disappears until it's valid code again. New objects sometimes don't appear until I collapse and re-expand the network. Sometimes (actually almost always) the network does not refresh when I finished typing and the editor keeps displaying a syntax error.

**Comment by tcstewar** [comment]
> How do I save anything?

It's actually automatically saving to file immediately.  I don't like that as a long-term solution, but it means that it can use the same reloading code as when you're using an external editor.

**Comment by jgosmann** [comment]
No, it's not. When I change it and then load it another browser it's still without the change.

**Comment by tcstewar** [comment]
> If I type to add ensemble, stuff I'm not editing disappears until it's valid code again.

Oops, I was having it do a partial reload when there's an Exception...  that meant that everything after the Exception disappeared.  That should be fixed now.

**Comment by tcstewar** [comment]
> No, it's not. When I change it and then load it another browser it's still without the change.

Yes it is.  :)  If you look at the file itself you can see it has changed.  But, you're quite right that there's a bug that if you open up a new browser window, it send you the old version of the file rather than reloading it.

**Comment by tcstewar** [comment]
> It would be cool to be able to create a new file from within the GUI.

Good plan... I think the same approach as in ```nengo_gui``` would work here: just edit the name in the toolbar.

**Comment by tcstewar** [comment]
> New objects sometimes don't appear until I collapse and re-expand the network

I think this is working better now.  I still have a bit of a problem if someone changes the uid (not the label) on an existing expanded Network (that requires a collapse and a re-open), but that's a pretty rare action.

**Comment by tcstewar** [comment]
> Sometimes (actually almost always) the network does not refresh when I finished typing and the editor keeps displaying a syntax error.

Hmm..  I'm not seeing that behaviour at all.  :(  Are there any weird errors shown in the console?

**Comment by jgosmann** [comment]
No really ... on the error which is displayed in the editor:
```
Traceback (most recent call last):
  File "/Volumes/Home/blubb/Documents/projects/uni/nengo_viz/nengo_viz/components/netgraph.py", line 52, in reload
    exec(code, locals)
  File "<string>", line 21, in <module>
NameError: name 'nen' is not defined
Traceback (most recent call last):
  File "/Volumes/Home/blubb/Documents/projects/uni/nengo_viz/nengo_viz/components/netgraph.py", line 52, in reload
    exec(code, locals)
  File "<string>", line 21, in <module>
AttributeError: 'module' object has no attribute 'Ens'
Traceback (most recent call last):
  File "/Volumes/Home/blubb/Documents/projects/uni/nengo_viz/nengo_viz/components/netgraph.py", line 52, in reload
    exec(code, locals)
  File "<string>", line 23
    with nengo.Network() as subsubnet:
       ^
SyntaxError: invalid syntax
```

**Comment by tcstewar** [comment]
The code editor's a fair bit better now, if people want to take a look:
- there's now a save button (and it only saves to file when you click it)
- no longer clearing the whole rest of the model due to syntax errors
- more syntax errors caught
- more robust updating of items as they change while you type
- removal of graphs on things that no longer exist due to edits

One big thing still to add is changing the file name, but I'm pretty happy with that.

**Comment by celiasmith** [comment]
Here's a fun one:
If you create an ensemble named 'ens', then do it again, the 2nd one takes the connections from the first.  If you then rename the 2nd one to, say ens1, you get this orphaned ensemble 'ens' that's not in the code but is in the netgraph.

**Comment by celiasmith** [comment]
(The problem I just posted will happen lots if you start copy/pasting).  This version is definitely much more robust!  I did find that on copy/paste of a connection, the pasted one doesn't work/exist often.  Especially on a 2nd paste.

**Comment by celiasmith** [comment]
A few times, hitting reload results in a new network drawn on top of an old one, so you have 2 copies.  Sometimes another reload fixes this, sometimes it doesn't.

**Comment by celiasmith** [comment]
Starting with 2d.py, I copy/paste the connection and change the first item to 'ens', and it doesn't draw a recurrent connection (it adds an arrow it seems).  on refresh it fixes itself though.

**Comment by tcstewar** [comment]
> A few times, hitting reload results in a new network drawn on top of an old one, so you have 2 copies.

I've seen this one a few times as well, but haven't figure out a way to replicate it.  If you figure out a way, let me know!

**Comment by jgosmann** [comment]
* If I use the reset button before saving the file all changes are lost. (And the confirmation dialog does not warn about that.)
* I added an ensemble which reused a variable name and then removed it. The result was that the removed ensemble did not disappear from the netgraph and kept it's connections while the remaining ensemble was there without its connections.

**Comment by tcstewar** [comment]
> I added an ensemble which reused a variable name and then removed it. The result was that the removed ensemble did not disappear from the netgraph and kept it's connections while the remaining ensemble was there without its connections.

That should be working now.  :)  Thanks for finding that!

**Comment by tcstewar** [comment]
> Starting with 2d.py, I copy/paste the connection and change the first item to 'ens', and it doesn't draw a recurrent connection (it adds an arrow it seems). on refresh it fixes itself though.

That should also be working now (basically, the connections we're updating whether they were recurrent or not when being moved around)

**Comment by cchan987** [comment]
The current save icon feels like a download icon to me. I think glyphicon glyphicon-floppy-save might be better?

**Comment by tbekolay** [comment]
Just tried it out, this looks pretty slick! I didn't do any rigorous testing of it though.

Only thing that I'm not a fan of is the text size. Especially for teaching tutorials, this is going to be way too small. Any chance we can make this configurable?

File: nengo_viz/config.py [hunk 1]
```diff
@@ -47,6 +47,7 @@ def dumps(self, uids):
                 lines.append('%s = %s' % (uid, obj.code_python(uids)))
                 for k in obj.config_params.keys():
                     v = getattr(self[obj], k)
+                    ##print(uid,k,v, 'from dumps')
```
Comment (tbekolay) [inline nengo_viz/config.py]: Probably remove this before merging?

File: nengo_viz/static/viz_ace.css [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#editor {
+	position: fixed;
```
Comment (tbekolay) [inline nengo_viz/static/viz_ace.css]: There's a tab here where there should be 4 spaces.

File: nengo_viz/templates/page.html [hunk 3]
```diff
@@ -11,15 +11,15 @@
...
         <link href="static//lib/css/bootstrap.min.css" rel="stylesheet" type="text/css" />
         <link href="static/viz_top_toolbar.css" rel="stylesheet" type="text/css" />
         <link href="static/lib/css/bootstrap.min.css" rel="stylesheet" type="text/css" />
         <link href="static/lib/js/jqueryFileTree/jqueryFileTree.css" rel="stylesheet" type="text/css">
     </head>
-
     <body onload="onload();">
-        <div id="top_toolbar">
-            <ul class="nav nav-pills">
+        <div id="top_toolbar_div">
```
Comment (tbekolay) [inline nengo_viz/templates/page.html]: Why make this change? Seems overly verbose...

Comment (cchan987) [inline nengo_viz/templates/page.html]: > Why make this change? Seems overly verbose...

I believe that this was a result of some debugging/testing where there was also a top_toolbar id in existence.

**Comment by tcstewar** [comment]
> Any chance we can make this configurable?

I totally agree.  I'm tempted to just toss a font++ button and a font-- button in the toolbar for now, and we can put it in a nice config system later.

**Comment by tcstewar** [comment]
> The current save icon feels like a download icon to me. I think glyphicon glyphicon-floppy-save might be better?

possibly, but does anyone still know what a floppy disc is?  ;)

**Comment by tbekolay** [comment]
> toss a font++ button and a font-- button in the toolbar for now

Works for me!

> does anyone still know what a floppy disc is?

No, but everyone knows that this is synonymous with save, for sure ;)

**Comment by cchan987** [comment]
> possibly, but does anyone still know what a floppy disc is? ;)

LOL! 

The (floppy) icon nowadays represents the concept of saving more than it represents floppy disks anyway. Via http://stackoverflow.com/questions/1019573/save-icon-still-a-floppy-disk

**Comment by hunse** [comment]
We should have a :cloud: for save. Get with the times, man! Also, how awesome would it be if we actually had a button to sync models with a central repository, and a way to search that repository all from within the GUI?

**Comment by tcstewar** [comment]
> The (floppy) icon nowadays represents the concept of saving more than it represents floppy disks anyway.

I suppose....  I've switched it over now....

**Comment by tcstewar** [comment]
> I totally agree. I'm tempted to just toss a font++ button and a font-- button in the toolbar for now, and we can put it in a nice config system later.

There are now font increase and decrease buttons.  I'm not happy with the icons for them, but it's fairly clear what they do....

File: nengo_viz/static/viz_ace.css (hunk 2 above)
Comment (tcstewar) [inline nengo_viz/static/viz_ace.css]: Fixed

File: nengo_viz/config.py (hunk 1 above)
Comment (tcstewar) [inline nengo_viz/config.py]: removed

**Comment by tcstewar** [comment]
> Also, how awesome would it be if we actually had a button to sync models with a central repository, and a way to search that repository all from within the GUI?

That would be some sweet github integration....  :)  And probably pretty doable, too.....

**Comment by celiasmith** [comment]
So I've been trying to come up with a simple reproduction for the 'duplicate on reload' issue.  I can't get a set number of steps to do it, but this general pattern gets it pretty much every time at some point.
1. Start with 2d.py
2. Copy paste the ensemble; then change the name to 'ens2'
3. Copy/paste the connection, change it to ens, ens
4. repeat step 3, copying the first connection (stimulus, ens), and changing it to various valid other connections

Hope that helps... I'll let you know if I have something better.  It happens quite frequently, whatever it is.

**Comment by tcstewar** [comment]
> So I've been trying to come up with a simple reproduction for the 'duplicate on reload' issue. I can't get a set number of steps to do it, but this general pattern gets it pretty much every time at some point.

Hmm..  that set of steps doesn't seem to include a reload anywhere.  Are you finding this problem happening without a reload?

**Comment by celiasmith** [comment]
No, what usually happens is that one of the connections doesn't appear, so i hit refresh and get the duplication.  (I was just doing this, and ran into problems deleting connections as well).

**Comment by tcstewar** [comment]
> No, what usually happens is that one of the connections doesn't appear, so i hit refresh and get the duplication. (I was just doing this, and ran into problems deleting connections as well).

Is that "connections don't appear" thing still happening for you given the recent changes I made?

**Comment by celiasmith** [comment]
Yes, this is the most recent version on the repo.  It seems to be some kind of list management issue, as I start getting errors like "ERROR finding label <Node (unlabeled) at 0x106e34610>" and "ValueError: list.remove(x): x not in list"  when doing this.

**Comment by tcstewar** [comment]
> Yes, this is the most recent version on the repo. It seems to be some kind of list management issue, as I start getting errors like "ERROR finding label " and "ValueError: list.remove(x): x not in list" when doing this.

Do you get those errors before or after you do a reload?

**Comment by tcstewar** [comment]
> Yes, this is the most recent version on the repo. It seems to be some kind of list management issue, as I start getting errors like "ERROR finding label " and "ValueError: list.remove(x): x not in list" when doing this.

And could you post the full error message?

**Comment by celiasmith** [comment]
I think those were after a reload (or two)... here's the full console from that session (of me following that sort of 'algorithm' i described):

Chriss-MacBook-Pro-6:examples celiasmi$ python 2d.py
('error parsing config', '_viz_config[ens2].pos=(0.7172032818933082, 0.2525035765379114)\n', NameError("name 'ens2' is not defined",))
('error parsing config', '_viz_config[ens2].size=(0.09433962264150943, 0.14492753623188406)\n', NameError("name 'ens2' is not defined",))
Starting nengo_viz server at http://localhost:8080
('error parsing config', '_viz_config[ens2].pos=(0.7172032818933082, 0.2525035765379114)\n', NameError("name 'ens2' is not defined",))
('error parsing config', '_viz_config[ens2].size=(0.09433962264150943, 0.14492753623188406)\n', NameError("name 'ens2' is not defined",))
ERROR finding label <Ensemble (unlabeled) at 0x106dffa50>
ERROR finding label <Ensemble (unlabeled) at 0x106dffad0>
ERROR finding label <Node (unlabeled) at 0x106dff050>
ERROR finding label <Ensemble (unlabeled) at 0x106de2950>
ERROR finding label <Ensemble (unlabeled) at 0x106de2490>
ERROR finding label <Node (unlabeled) at 0x106de2410>
Traceback (most recent call last):
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/swi.py", line 227, in handle_ws_request
    getattr(self, command)(*[client] + args[1:], **db)
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/server.py", line 143, in ws_viz_component
    component.finish()
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/components/sim_control.py", line 40, in finish
    self.viz.finish()
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/viz.py", line 98, in finish
    self.viz.remove_sim(self)
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/viz.py", line 329, in remove_sim
    self.viz_sims.remove(viz_sim)
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/swi.py", line 227, in handle_ws_request
    getattr(self, command)(*[client] + args[1:], **db)
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/server.py", line 143, in ws_viz_component
    component.finish()
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/components/sim_control.py", line 40, in finish
    self.viz.finish()
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/viz.py", line 98, in finish
    self.viz.remove_sim(self)
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/viz.py", line 329, in remove_sim
    self.viz_sims.remove(viz_sim)
ValueError: list.remove(x): x not in list
ERROR finding label <Ensemble (unlabeled) at 0x106e5d450>
ERROR finding label <Node (unlabeled) at 0x106e34610>
ERROR finding label <Ensemble (unlabeled) at 0x106de2110>
Traceback (most recent call last):
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/swi.py", line 227, in handle_ws_request
    getattr(self, command)(*[client] + args[1:], **db)
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/server.py", line 143, in ws_viz_component
    component.finish()
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/components/sim_control.py", line 40, in finish
    self.viz.finish()
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/viz.py", line 98, in finish
    self.viz.remove_sim(self)
  File "/Users/celiasmi/Documents/nengo/nengo_viz/nengo_viz/viz.py", line 329, in remove_sim
    self.viz_sims.remove(viz_sim)
ValueError: list.remove(x): x not in list
^C

**Comment by celiasmith** [comment]
I should note, the stuff at the beginning was from a bad .cfg file; I deleted it and started fresh and still have similar problems.

**Comment by tcstewar** [comment]
> I think those were after a reload (or two)...

Have you ever seen this problem happen without a reload?

I'm more and more convinced that's where the problem is.

Also, do you mean an actual browser reload or do you mean hitting the reset button?

**Comment by celiasmith** [comment]
I mean hitting the browser 'refresh' button.  It definitely often happens after a reload... so you might be right.  I do feel i've seen it without reloading though... i'll try to verify.

**Comment by tcstewar** [comment]
I just rebased this as it was getting a bit out of date...

**Comment by jgosmann** [comment]
For the font size icons, I would prefer to use `glyphicon glyphicon-zoom-in` and `glyphicon glyphicon-zoom-out` (you could mistake those for zooming the net graph, but at least they indicate that they zoom something, whereas those arrows are pretty unclear)

If you'd want to get fancy take `glyphicon glyphicon-font` for both icons and scale it differently.

**Comment by tbekolay** [comment]
- [ ] Make sure there's a keyboard shortcut to pop up / close the editor (from #269)

**Comment by jgosmann** [comment]
:crystal_ball: I foresee issue with the IPython integration.

**Comment by tcstewar** [comment]
> I foresee issue with the IPython integration.

Yup, I'm doing that merge now....   my plan is to use your ```interactive``` variable and just not have the code editor appear.

**Comment by tcstewar** [comment]
The code editor now doesn't appear when run in IPython.  And we are rebased again.  :)

**Comment by celiasmith** [comment]
Replication is your friend.  This works every time on my machine:

1. Load 2d.py
2. Copy/paste the ensemble and rename to 'ens2'
3. Hit reload
4. Copy/paste the connection
5. Hit reload

Voila, you've got duplicates (you might have to move the existing ensembles to see the duplicates).

**Comment by tbekolay** [comment]
This is in a pretty good state; can use some work still, but I'm merging it in now 'cause I gotta run soon and we need to release!

**Comment by s72sue** [comment]
How does the code editor handle spacebar?

I mean if I am typing and using spacebar in the code editor, are we
preventing it from playing/pausing the simulation?

**Comment by jgosmann** [comment]
oha, this has to be fixed too ...

**Comment by jgosmann** [comment]
It's basically unusable now

**Comment by tbekolay** [comment]
The code editor, when it has focus, takes all key events.

**Comment by jgosmann** [comment]
no it does not, at least not in Firefox

**Comment by jgosmann** [comment]
I think, the propagation of the events has to be stopped

**Comment by jgosmann** [comment]
same in Chrome

**Comment by s72sue** [comment]
Yes I tried that too, it doesn't take all the key events in Chrome (on a
Windows OS)

**Comment by tcstewar** [comment]
I've submitted a fix for this: #312

**Comment by tcstewar** [comment]
(note: the plan is to keep testing over the weekend and do a 0.1.1 release on sunday night)

**Comment by jgosmann** [comment]
ok, I'm creating milestone for that, just found another bug

**Comment by jgosmann** [comment]
Another rebase that went wrong. [This change shouldn't be there.](https://github.com/nengo/nengo_gui/commit/7875a26dde2137e0b6402dc0dc5677b26f18bc4f#diff-b161a997dd41bda62069ee5a3e52a606L20)

**Comment by jgosmann** [comment]
Fix in #316

File: nengo_viz/static/viz_ace.css [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#editor {
+    position: fixed;
```
Comment (jgosmann) [inline nengo_viz/static/viz_ace.css]: I'm a bit surprised about the amount of elements with `position: fixed` in the code base. I had to change quite a number of those to `absolute` (or sometimes to `relative`). I wouldn't be surprised if this were another case where it should be `absolute`.

**Pull request merged**